File: k9mail/src/main/java/com/fsck/k9/pEp/PEpPermissionChecker.java
package com.fsck.k9.pEp;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;

import androidx.core.content.ContextCompat;

import security.pEp.permissions.PermissionChecker;

public class PEpPermissionChecker implements PermissionChecker {

    private final Context context;

    public PEpPermissionChecker(Context context) {
        this.context = context;
    }

    public static Boolean hasWriteExternalPermission(Context context) {
        int res = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE);
        return (res == PackageManager.PERMISSION_GRANTED);
    }

    public static Boolean hasWriteContactsPermission(Context context) {
        int res = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_CONTACTS);
        return (res == PackageManager.PERMISSION_GRANTED);
    }

    @Override
    public boolean hasBasicPermission() {
        return hasContactsPermission() && hasWriteExternalPermission();
    }

    @Override
    public boolean hasWriteExternalPermission() {
        int res = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE);
        return (res == PackageManager.PERMISSION_GRANTED);
    }

    @Override
    public boolean doesntHaveWriteExternalPermission() {
        int res = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE);
        return (res == PackageManager.PERMISSION_DENIED);
    }

    @Override
    public boolean doesntHaveContactsPermission() {
        int readRes = ContextCompat.checkSelfPermission(context, Manifest.permission.READ_CONTACTS);
        int writeRes = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_CONTACTS);
        return (readRes == PackageManager.PERMISSION_DENIED &&
                writeRes == PackageManager.PERMISSION_DENIED
        );
    }

    @Override
    public boolean hasContactsPermission() {
        int readRes = ContextCompat.checkSelfPermission(context, Manifest.permission.READ_CONTACTS);
        int writeRes = ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_CONTACTS);
        return (readRes == PackageManager.PERMISSION_GRANTED &&
                writeRes == PackageManager.PERMISSION_GRANTED
        );
    }


}

File: k9mail/src/main/java/security/pEp/ui/permissions/PEpPermissionChecker.kt
package security.pEp.ui.permissions

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RequiresApi

import androidx.core.content.ContextCompat

import security.pEp.permissions.PermissionChecker

class PEpPermissionChecker(private val context: Context) : PermissionChecker {

    override fun hasBasicPermission(): Boolean {
        return hasContactsPermission() && hasWriteExternalPermission()
    }

    override fun hasWriteExternalPermission(): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) return true
        val res = getExternalStoragePermission()
        return arePermissionsGranted(res)
    }

    override fun doesntHaveWriteExternalPermission(): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) return false
        val res = getExternalStoragePermission()
        return arePermissionsDenied(res)
    }

    override fun doesntHaveContactsPermission(): Boolean {
        val readRes = getReadContactsPermission()
        val writeRes = getWriteContactsPermission()
        return arePermissionsDenied(readRes, writeRes)
    }

    override fun hasContactsPermission(): Boolean {
        val readRes = getReadContactsPermission()
        val writeRes = getWriteContactsPermission()
        return arePermissionsGranted(readRes, writeRes)
    }

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    override fun doesntHavePostNotificationsPermission(): Boolean {
        val res = getPostNotificationsPermission()
        return arePermissionsDenied(res)
    }

    private fun arePermissionsGranted(vararg permissionsStatusResult: Int): Boolean {
        return permissionsStatusResult
                .map { permissionStatus -> permissionStatus == PackageManager.PERMISSION_GRANTED }
                .reduce { permissionsGranted, newValue -> permissionsGranted && newValue }
    }

    private fun arePermissionsDenied(vararg permissionsStatusResult: Int): Boolean {
        return permissionsStatusResult
                .map { permissionStatus -> permissionStatus == PackageManager.PERMISSION_DENIED }
                .reduce { permissionsDenied, newValue -> permissionsDenied && newValue }
    }

    private fun getReadContactsPermission(): Int {
        return ContextCompat.checkSelfPermission(context, Manifest.permission.READ_CONTACTS)
    }

    private fun getWriteContactsPermission(): Int {
        return ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_CONTACTS)
    }

    private fun getExternalStoragePermission(): Int {
        return ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE)
    }

    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
    private fun getPostNotificationsPermission(): Int {
        return ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS)
    }
}
